packages/circuit-ui/components/TextArea/useAutoExpand.ts

```diff
@@ -0,0 +1,77 @@
+/* eslint-disable no-param-reassign */
+/**
+ * Copyright 2021, SumUp Ltd.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FormEvent, RefObject, useCallback, useEffect } from 'react';
+
+import { HTMLCircuitInputElement } from '../Input/Input';
+import { useComponentSize } from '../../hooks/useComponentSize';
+
+import { TextAreaProps } from '.';
+
+type ModifiedProps = Omit<TextAreaProps, 'minRows' | 'rows'> & {
+  rows: TextAreaProps['minRows'];
+};
+
+export const useAutoExpand = (
+  ref: RefObject<HTMLCircuitInputElement>,
+  { minRows, rows = minRows, onInput, ...props }: TextAreaProps,
+): ModifiedProps => {
+  const autoExpand = rows === 'auto';
+
+  const { height, width } = useComponentSize(ref) as DOMRect;
+
+  const updateElementHeight = useCallback(
+    (el: EventTarget & HTMLTextAreaElement) => {
+      const previousValue = el.value ?? '';
+      const shouldUsePlaceholder = !previousValue.length && !!props.placeholder;
+      if (shouldUsePlaceholder) {
+        el.value = props.placeholder ?? '';
+      }
+
+      el.style.height = 'auto';
+      if (el.scrollHeight) {
+        el.style.height = `${el.scrollHeight}px`;
+      }
+      if (shouldUsePlaceholder) {
+        el.value = previousValue;
+      }
+    },
+    [props.placeholder],
+  );
+
+  const inputHandler = useCallback(
+    (e: FormEvent<HTMLCircuitInputElement>) => {
+      updateElementHeight(e.currentTarget);
+      onInput?.(e);
```

Comment:
I'm not sure what you mean.

what I know is that for the autosize to work seamlessly, I need to call that handler after *each and every* alterations; therefore, I need onInput instead of onChange.

for the typical use, as a component's library, I prefere not to interfere with how the user use the component.
If it defined the `onInput` I call it there.

But as I see the code, I think we can improve a little bit this handler,
by calling updateElementHeight after the passed handler. And only if the event is not defaultPrevented.

 - [ ] avoid updateElementHeight calls if event is prevented.

Reply:
We might have to discuss that further, as I see other issues that could be related.

https://github.com/sumup-oss/circuit-ui/issues/1025